src/Distributed/halo_communication.jl

```diff
@@ -0,0 +1,165 @@
+end
+
+#####
+##### fill_east_and_west_halos!   }
+##### fill_north_and_south_halos! } for non-communicating boundary conditions (fallback)
+##### fill_top_and_bottom_halos!  }
+#####
+
+for (side, opposite_side) in zip([:east, :north, :top], [:west, :south, :bottom])
+    fill_both_halos! = Symbol("fill_$(side)_and_$(opposite_side)_halos!")
+    fill_side_halo! = Symbol("fill_$(side)_halo!")
+    fill_opposite_side_halo! = Symbol("fill_$(opposite_side)_halo!")
+
+    @eval begin
+        function $fill_both_halos!(c, bc_side, bc_opposite_side, arch, barrier, grid, args...)
+            event_side = $fill_side_halo!(c, bc_side, child_architecture(arch), barrier, grid, args...)
+            event_opposite_side = $fill_opposite_side_halo!(c, bc_opposite_side, child_architecture(arch), barrier, grid, args...)
+            return event_side, event_opposite_side
+        end
+    end
+end
+
+#####
+##### fill_east_and_west_halos!   }
+##### fill_north_and_south_halos! } for when both halos are communicative
+##### fill_top_and_bottom_halos!  }
+#####
+
+for (side, opposite_side) in zip([:east, :north, :top], [:west, :south, :bottom])
+    fill_both_halos! = Symbol("fill_$(side)_and_$(opposite_side)_halos!")
+    send_side_halo = Symbol("send_$(side)_halo")
+    send_opposite_side_halo = Symbol("send_$(opposite_side)_halo")
+    recv_and_fill_side_halo! = Symbol("recv_and_fill_$(side)_halo!")
+    recv_and_fill_opposite_side_halo! = Symbol("recv_and_fill_$(opposite_side)_halo!")
+
+    @eval begin
+        function $fill_both_halos!(c, bc_side::HaloCommunicationBC, bc_opposite_side::HaloCommunicationBC, arch, barrier, grid, c_location, args...)
+            @assert bc_side.condition.from == bc_opposite_side.condition.from  # Extra protection in case of bugs
+            my_rank = bc_side.condition.from
+
+            $send_side_halo(c, grid, c_location, my_rank, bc_side.condition.to)
+            $send_opposite_side_halo(c, grid, c_location, my_rank, bc_opposite_side.condition.to)
+
+            $recv_and_fill_side_halo!(c, grid, c_location, my_rank, bc_side.condition.to)
+            $recv_and_fill_opposite_side_halo!(c, grid, c_location, my_rank, bc_opposite_side.condition.to)
+
+            return nothing, nothing
+        end
+    end
+end
+
+#####
+##### Sending halos
+#####
+
+for side in sides
+    side_str = string(side)
+    send_side_halo = Symbol("send_$(side)_halo")
+    underlying_side_boundary = Symbol("underlying_$(side)_boundary")
+    side_send_tag = Symbol("$(side)_send_tag")
+
+    @eval begin
+        function $send_side_halo(c, grid, c_location, my_rank, rank_to_send_to)
+            send_buffer = $underlying_side_boundary(c, grid, c_location)
+            send_tag = $side_send_tag(my_rank, rank_to_send_to)
+
+            @debug "Sending " * $side_str * " halo: my_rank=$my_rank, rank_to_send_to=$rank_to_send_to, send_tag=$send_tag"
+            status = MPI.Isend(send_buffer, rank_to_send_to, send_tag, MPI.COMM_WORLD)
```

Comment:
Do you ever do an MPI_Wait on the status? Officially you are supposed to do that at some point I think?

Reply:
Ah no I thought `MPI.Recv!` can't do anything until the `MPI.Isend` is finished, but an `MPI.Wait!` or `MPI.Waitall!` would be good to add. I can add those.

Sounds like it could save us from some obscure bugs in the future.